resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/ReactivePersistenceLayer.java

```diff
@@ -225,23 +244,21 @@ private static void traverseParentsByResource(final Resource resource, final BiP
      * @param resource the top-level resource to get the leaves of
      * @return a {@link Stream} of {@link LeafResource}s from the resource repository persistence store
      */
-    private FunctionalIterator<LeafResource> collectLeaves(final Resource resource) {
+    private Source<LeafResource, NotUsed> collectLeaves(final Resource resource) {
         if (resource instanceof ParentResource) {
             // Treat resource as a ParentResource
             ParentResource parentResource = (ParentResource) resource;
             // Get the children
-            FunctionalIterator<ResourceEntity> childEntities = resourceRepository.iterateFindAllByParentId(parentResource.getId());
-            FunctionalIterator<Resource> childResources = childEntities
-                    .map(ResourceEntity::getResource);
-            // Recurse over all further children
-            return childResources.flatMap(this::collectLeaves)
+            return resourceRepository.streamFindAllByParentId(parentResource.getId())
+                    .map(ResourceEntity::getResource)
+                    // Recurse over further children
+                    .flatMapConcat(this::collectLeaves)
                     .map(leafResource -> resolveParentsUpto(leafResource, resource));
-        } else {
+        } else if (resource instanceof LeafResource) {
             // If we have reached a leaf, then done
-            // Return one-element iterator (this could probably lead to some performance penalty)
-            // It should be a negligible hit to performance, relative how much other processing happens per-leaf
-            return FunctionalIterator.fromIterator(
-                    Collections.singletonList((LeafResource) resource).iterator());
+            return Source.single((LeafResource) resource);
+        } else {
+            throw new UnknownResourceTypeException(String.format("Resource %s is neither Parent nor Leaf", resource));
```

Comment:
Maybe worth an error message here just in case
```suggestion
            LOGGER.error("Invalid class {}", resource.getClassName());
            throw new UnknownResourceTypeException(String.format("Resource %s is neither Parent nor Leaf", resource));
```

Reply:
Added the error log message